### PR TITLE
[DAG] Add DAG Visualization with Jupyter Support

### LINKED ISSUE #4277: Add Basic Visualization Support for DAGs
Implement a simple DAG visualization feature to help users view task dependencies and execution flow. This visualization will make it easier to understand and debug complex DAG structures.

**Proposed Solution:**

1. Add a `plot` method in `Dag`:
   - Leverage `pydot` and `graphviz` for rendering the DAG.
   - Save the generated visualization as an image file (e.g., `.png`).
   - Display the image inline in Jupyter Notebooks if available.

2. Key implementation steps:
   - Extract nodes and edges using DFS to ensure unique node names.
   - Build the visualization graph by adding nodes and edges.
   - Provide optional file path for saving the visualization.

**Requirements:**

- Check for `pydot` and `graphviz` installation before plotting, with clear installation instructions if missing.

**References:**

[Ray’s implementation of DAG visualization](https://github.com/ray-project/ray/blob/master/python/ray/dag/vis_utils.py) could serve as a useful reference for structuring this feature.

### PR DESCRIPTION
**Pull request by andylizf**
Resolve #4277.

Adds a `plot` method to the `Dag` class to visualize task dependencies. Automatically displays inline in Jupyter or saves to a file in other environments.

#### Example:
```python
from sky import Dag, Task

with Dag() as dag:
    task_a = Task(name="A")
    task_b = Task(name="B")
    task_c = Task(name="C")
    task_a >> task_b >> task_c

dag.plot()
```

**Notes:**
- Requires `matplotlib` and `networkx`.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
    - Tested in Jupyter Notebook (`.ipynb`)
    - Tested in regular Python environment
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

**Comment by andylizf** [comment]
@cblmemo PTAL, thanks!

File: sky/dag.py [hunk 1]
```diff
@@ -258,6 +259,79 @@ def is_connected_dag(self) -> bool:
...
+                                horizontalalignment='center',
+                                bbox=dict(facecolor='white',
+                                          edgecolor='gray',
+                                          boxstyle='round,pad=0.3'))
+
+        ax.margins(0.2)
+        plt.subplots_adjust(left=0.15, right=0.85, top=0.85, bottom=0.15)
+
+        if to_file is None:
+            tmp_file = tempfile.NamedTemporaryFile(suffix='.png', delete=False)
```
Comment (cblmemo) [inline sky/dag.py:316]: add format as arguments?

File: sky/dag.py [hunk 2]
```diff
@@ -258,6 +259,79 @@ def is_connected_dag(self) -> bool:
...
+
+        plt.savefig(to_file, bbox_inches='tight')
+        plt.close()
+
+        try:
+            # Try to display the image in Jupyter Notebook
+            # pylint: disable=import-outside-toplevel
+            from IPython.display import display
+            from IPython.display import Image
+            display(Image(filename=to_file))
```
Comment (cblmemo) [inline sky/dag.py:328]:
```suggestion
            from IPython import display
            display.display(display.Image(filename=to_file))
```

**Review by cblmemo — APPROVED** [review]
Thanks! LGTM. Feel free to merge it

**Comment by andylizf** [comment]
Hi @cblmemo, could you handle the merge for me?

**Pull request merged**